File: src/walkway_navigation/src/patrol_bot.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "nav2_msgs/action/navigate_to_pose.hpp"

class PatrolBot : public rclcpp::Node {
    public:
    using NavigateToPose = nav2_msgs::action::NavigateToPose;

    PatrolBot() : Node("patrol_bot") {
        // 1. Create the Action Client
        client_ptr_ = rclcpp_action::create_client<NavigateToPose>(this, "navigate_to_pose");
    }

    void send_goal(double x, double y, double theta_w) {
        // 2. Wait for Nav2 to be ready
        if (!client_ptr_->wait_for_action_server(std::chrono::seconds(10))) {
            RCLCPP_ERROR(this->get_logger(), "Nav2 Action server not available!");
            return;
        }

        // 3. Construct the Goal Message
        auto goal_msg = NavigateToPose::Goal();
        goal_msg.pose.header.frame_id = "map";
        goal_msg.pose.pose.position.x = x;
        goal_msg.pose.pose.position.y = y;
        goal_msg.pose.pose.orientation.w = theta_w;

        // 4. Send asynchronously
        RCLCPP_INFO(this->get_logger(), "Sending goal!");
        client_ptr_->async_send_goal(goal_msg);
    }

    private:
    rclcpp_action::Client<NavigateToPose>::SharedPtr client_ptr_;
};

int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PatrolBot>();
    node->send_goal(2.0, 2.0, 1.0);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/walkway_navigation/scripts/patrol_bot.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from nav2_msgs.action import NavigateToPose

class PatrolBot(Node):
    def __init__(self):
        super().__init__('patrol_bot')
        # 1. Create an Action Client targeting the Nav2 Server
        self.action_client = ActionClient(self, NavigateToPose, 'navigate_to_pose')

    def send_goal(self, x, y, theta_w):
        # 2. Construct the Goal Message
        goal_msg = NavigateToPose.Goal()
        goal_msg.pose.header.frame_id = 'map'
        goal_msg.pose.pose.position.x = x
        goal_msg.pose.pose.position.y = y 
        goal_msg.pose.pose.orientation.w = theta_w

        # 3. Wait for Nav2 to be ready, then send asynchronously
        self.get_logger().info('Waiting for Nav2 server...')
        self.action_client.wait_for_server()
        self.get_logger().info('Sending goal!')
        self.action_client.send_goal_async(goal_msg)

def main(args=None):
    rclpy.init(args=args)
    node = PatrolBot()
    # Command the robot to drive to coordinate (2.0, 2.0)
    node.send_goal(2.0, 2.0, 1.0)
    rclpy.spin(node)
    rclpy.shutdown()
